beam-core/Database/Beam/Schema/Indices.hs

```diff
@@ -0,0 +1,357 @@
+
+    Although this all is pretty cumbersome and scaring at a first glance, and dependencies on
+    'hlist' and 'vinyl' which may be required here are quite heavyweight.
+-}
+
+-- | Generic helper for 'AutoTableIndices'.
+class GAutoTableIndices (x :: * -> *) where
+    -- | Returns list of deferred indices.
+    --   Exactly this type is required, later in migrations knowing that list size does
+    --   not depend on names (only on structure of the database) is important.
+    autoTableIndices' :: DList (x p -> TableIndex)
+
+instance GAutoTableIndices x => GAutoTableIndices (M1 i f x) where
+    autoTableIndices' = contramapTableIndicesBuilder unM1 $ autoTableIndices' @x
+
+instance (GAutoTableIndices x, GAutoTableIndices y) =>
+          GAutoTableIndices (x :*: y) where
+    autoTableIndices' =
+        contramapTableIndicesBuilder (\(x :*: _) -> x) (autoTableIndices' @x) <>
+        contramapTableIndicesBuilder (\(_ :*: y) -> y) (autoTableIndices' @y)
+
+instance GAutoTableIndices (Rec0 x) where
+    autoTableIndices' = mempty
+
+instance {-# OVERLAPPING #-}
+         (Beamable (PrimaryKey tbl'), IndexFromReference (tbl :-> tbl')) =>
+         GAutoTableIndices (Rec0 (PrimaryKey tbl' (TableField tbl))) where
+    autoTableIndices' =
+        if tableValuesNeeded (Proxy @(PrimaryKey tbl')) == 0
+        then DL.empty
+        else DL.singleton $ \(K1 referringField) ->
+            let pkFields = allBeamValues
+                              (\(Columnar' (TableField fieldNm)) -> fieldNm)
+                              referringField
+                opts = referenceIndexOptions (Proxy @(tbl :-> tbl'))
+                -- unsafe call, but at this point we know the list is not empty
+            in TableIndex (fromList pkFields) opts
+
+instance {-# OVERLAPPING #-}
+         (Beamable (PrimaryKey tbl'), IndexFromReference (tbl :-> tbl')) =>
+         GAutoTableIndices (Rec0 (PrimaryKey tbl' (Nullable (TableField tbl)))) where
+    autoTableIndices' =
+        if tableValuesNeeded (Proxy @(PrimaryKey tbl')) == 0
+        then DL.empty
+        else DL.singleton $ \(K1 referringField) ->
+            let pkFields = allBeamValues
+                              (\(Columnar' (TableField fieldNm)) -> fieldNm)
+                              referringField
+                opts = referenceIndexOptions (Proxy @(tbl :-> tbl'))
+                -- unsafe call, but at this point we know the list is not empty
+            in TableIndex (fromList pkFields) opts
+
+-- | Traverses fields of the given table and builds indices for all encountered 'PrimaryKey's.
+class AutoEntityIndex be db tbl where
+    autoEntityIndices :: EntityIndices be db tbl
+
+-- Other types of entities are approaching, and we probably don't want to define
+-- instances for all of them.
+-- | Traverses the given table and for every field which is some 'PrimaryKey'
+-- makes corresponding SQL index, this allows "JOIN"s on this table perform nicely.
+instance {-# OVERLAPPABLE #-}
```

Comment:
Why are these instances overlappable?

Reply:
Well, I'm not sure whether is it a good choice, but I saw that there are plans to support many different entities type, while we want to have indices only for `TableEntities`. It looked like the one who will enable support for those other entities should not bother about indices at all, thus a default overlappable instance.
Should I define instances for each entity type explicitly instead?